File: shared_code_templates/stack_in_ruby/code.rb
class Node

  attr_accessor :nxt, :value

  def initialize(value)
    @value = value
    @nxt = nil
  end

end


class Stack

  attr_reader :head

  def initialize()
    @head = nil 
  end

  def push(value)
    node = Node.new(value)
    if @head != nil
      node.nxt = @head
    end
    @head = node
  end

  def pop
    if @head == nil
      return nil
    else
      old = @head
      @head = @head.nxt
      value = old.value
      # [tag:todo:gem] Need to delete "old"
      return value
    end
  end

end

File: shared_code_templates/stack_in_cc/code.cc
#include "code.h"

Stack::Stack() {
    head = NULL;
    return;
}

Stack::~Stack() {
    while (head) {
        Node *next = head->nxt;
        delete head;
        head = next;
    }
    return;
}

void Stack::push(int value) {
    Node *node = new Node;
    node->value = value;
    node->nxt = head;
    head = node;
    return;
}

int Stack::pop() {
    Node *popNode = head;
    if (head == NULL) {
        //[tag:todo:gem]
        //throw StackError(E_EMPTY);
        throw "stack is empty";
    }

    int value = head->value;
    head = head->nxt;
    delete popNode;
    return value;
}

//ogfileid:19
